### PR TITLE
Update streamlines formats example

### PR DESCRIPTION
**Pull request by frheault**
Considering the new StatefulTractogram and the fact that the old example only covered the Dpy format. This update now contain a more complete list of file format, the new API and new data that can be fetched.

### CODE REVIEW

File: dipy/io/utils.py [hunk 1]
```diff
@@ -1,6 +1,8 @@
 """ Utility functions for file formats """
 import logging
+import os
 
+import dipy
```
Comment (skoudoro) [inline dipy/io/utils.py]: Can you replace this by `from dipy.io.stateful_tractogram import StatefulTractogram`

File: dipy/io/utils.py [hunk 2]
```diff
@@ -177,6 +182,8 @@ def get_reference_info(reference):
     elif isinstance(reference, dict) and 'magic_number' in reference:
         header = reference
         is_trk = True
+    elif isinstance(reference, dipy.io.stateful_tractogram.StatefulTractogram):
```
Comment (skoudoro) [inline dipy/io/utils.py]: To follow up the comment above, Can you just use `StatefulTractogram` here

File: doc/examples/streamline_formats.py [hunk 3]
```diff
@@ -9,78 +9,223 @@
...
-dpw.write_tracks(streamlines)
+print(is_header_compatible(reference_anatomy, bundles_filename[0]))
 
 """
-Write one track
+If a TRK was generated with a valid header, but the reference NIFTI was lost
+an header can be generated to then generate a fake NIFTI file.
+
+If you wish to manually save Trk and Tck file using nibabel streamlines
+API for more freedom of action (not recommanded for beginner) you can
```
Comment (skoudoro) [inline doc/examples/streamline_formats.py]: typo, `recommended`

File: doc/examples/streamline_formats.py [hunk 4]
```diff
@@ -9,78 +9,223 @@
...
+affine, dimensions, voxel_sizes, voxel_order = cc_sft.space_attribute
+cc_density = density_map(cc_streamlines_vox, dimensions)
+laf_density = density_map(laf_streamlines_vox, dimensions)
+raf_density = density_map(raf_streamlines_vox, dimensions)
+lpt_density = density_map(lpt_streamlines_vox, dimensions)
+rpt_density = density_map(rpt_streamlines_vox, dimensions)
 
+"""
+Replacing streamlines is possible, but if the state was modified between
+operation such as this one is not recommanded:
```
Comment (skoudoro) [inline doc/examples/streamline_formats.py]: same as above `recommended`

File: doc/examples/streamline_formats.py [hunk 5]
```diff
@@ -9,78 +9,223 @@
...
+-> cc_sft.streamlines = cc_streamlines_vox
+
+It is recommanded to re-create a new StatefulTractogram object and
```
Comment (skoudoro) [inline doc/examples/streamline_formats.py]: typo `recommended`

File: doc/examples/streamline_formats.py [hunk 6]
```diff
@@ -9,78 +9,223 @@
...
+explicitely specified in which space the streamline are. Be careful to follow
```
Comment (skoudoro) [inline doc/examples/streamline_formats.py]: typo `explicitly` and `streamlines`

**Comment by skoudoro** [comment]
Thank you for the fix, I do not understand why Travis is failing, that's strange...

**Comment by frheault** [comment]
@skoudoro I fixed it, I just remember why I imported Dipy only instead of StatefulTractogram in dipy/io/utils.py. It was a cyclical important causing everything to crash.

I put it back, it is working now

**Comment by skoudoro** [comment]
> It was a cyclical important causing everything to crash.

wow, I really do not like that. That is not good, We really need to understand why it is so easy to create a cyclical import when we create a tutorial and correct that. I will put it up on my todo list

**Comment by skoudoro** [comment]
Maybe `get_reference_info` should go back to `dipy.io.stateful_tractogram`.

In any case, it will be on another PR.  Need to think more about it.

File: doc/examples/streamline_formats.py [hunk 7]
```diff
@@ -9,78 +9,223 @@
 
 DIPY_ can read and write many different file formats. In this example
 we give a short introduction on how to use it for loading or saving
-streamlines.
+streamlines. The new statefull tractogram class was made to reduce errors
+cause by spatial transformation and complex file format convention.
```
Comment (Garyfallidis) [inline doc/examples/streamline_formats.py]: caused by spatial transformations and complex file format conventions.

File: doc/examples/streamline_formats.py [hunk 8]
```diff
@@ -9,78 +9,223 @@
...
+laf_sft = load_tractogram(bundles_filename[1], reference_anatomy)
+raf_sft = load_tractogram(bundles_filename[3], reference_anatomy)
 
 """
-2. We also work on our HDF5 based file format which can read/write massive
-   datasets (as big as the size of your free disk space). With `Dpy` we can
-   support
+These files contain invalid streamlines (negative values once in voxel space)
+This is not considered a valid tractography file, but it is possible to load
+it anyway
```
Comment (Garyfallidis) [inline doc/examples/streamline_formats.py]: anyway.

File: doc/examples/streamline_formats.py [hunk 9]
```diff
@@ -9,78 +9,223 @@
...
+the same variety of input as ``get_reference_info``
 """
 
-dpw.write_tracks(streamlines)
+print(is_header_compatible(reference_anatomy, bundles_filename[0]))
 
 """
-Write one track
+If a TRK was generated with a valid header, but the reference NIFTI was lost
+an header can be generated to then generate a fake NIFTI file.
```
Comment (Garyfallidis) [inline doc/examples/streamline_formats.py]: a header

File: doc/examples/streamline_formats.py [hunk 10]
```diff
@@ -9,78 +9,223 @@
...
+
+If you wish to manually save Trk and Tck file using nibabel streamlines
+API for more freedom of action (not recommended for beginner) you can
```
Comment (Garyfallidis) [inline doc/examples/streamline_formats.py]: beginners

File: doc/examples/streamline_formats.py [hunk 11]
```diff
@@ -9,78 +9,223 @@
...
+create valid header using create_tractogram_header
```
Comment (Garyfallidis) [inline doc/examples/streamline_formats.py]: create a valid

File: doc/examples/streamline_formats.py [hunk 12]
```diff
@@ -9,78 +9,223 @@
...
+nifti_header = create_nifti_header(affine, dimensions, voxel_sizes)
+nib.save(nib.Nifti1Image(np.zeros(dimensions), affine, nifti_header),
+         'fake.nii.gz')
+nib.save(reference_anatomy, os.path.basename(ref_anat_filename))
 
 """
-or one track each time.
+Once loaded, no matter the original file format, the stateful tractogram is
+self contained and maintain a valid state and can be easily be manipulated
+Let's save all file as TRK to visualize in TrackVis for example.
```
Comment (Garyfallidis) [inline doc/examples/streamline_formats.py]: files

File: doc/examples/streamline_formats.py [hunk 13]
```diff
@@ -9,78 +9,223 @@
...
-dpw.write_track(streamlines[0])
+nifti_header = create_nifti_header(affine, dimensions, voxel_sizes)
+nib.save(nib.Nifti1Image(np.zeros(dimensions), affine, nifti_header),
+         'fake.nii.gz')
+nib.save(reference_anatomy, os.path.basename(ref_anat_filename))
 
 """
-or one track each time.
+Once loaded, no matter the original file format, the stateful tractogram is
+self contained and maintain a valid state and can be easily be manipulated
```
Comment (Garyfallidis) [inline doc/examples/streamline_formats.py]: maintains (too many ands?)

File: doc/examples/streamline_formats.py [hunk 14]
```diff
@@ -9,78 +9,223 @@
...
+Let's save all file as TRK to visualize in TrackVis for example.
+However, when loaded the lpt and rpt files contains invalid streamlines and
```
Comment (Garyfallidis) [inline doc/examples/streamline_formats.py]: contain

File: doc/examples/streamline_formats.py [hunk 15]
```diff
@@ -9,78 +9,223 @@
...
+print(lpt_sft.is_bbox_in_vox_valid())
+save_tractogram(lpt_sft, 'lpt.trk')
 
+print(rpt_sft.is_bbox_in_vox_valid())
+rpt_sft.remove_invalid_streamlines()
+print(rpt_sft.is_bbox_in_vox_valid())
+save_tractogram(rpt_sft, 'rpt.trk')
+
+"""
+Some functions in dipy require streamlines to be in voxel space so computation
```
Comment (Garyfallidis) [inline doc/examples/streamline_formats.py]: DIPY

File: doc/examples/streamline_formats.py [hunk 16]
```diff
@@ -9,78 +9,223 @@
...
+can be perfomed on a grid (connectivity matrix, ROIs masking, density map).
+The stateful tractogram class provided safe function for such manipulation.
```
Comment (Garyfallidis) [inline doc/examples/streamline_formats.py]: provides

Comment (Garyfallidis) [reply doc/examples/streamline_formats.py]: functions?

File: doc/examples/streamline_formats.py [hunk 17]
```diff
@@ -9,78 +9,223 @@
...
+These function can be called safely over and over, by knowing in which state
```
Comment (Garyfallidis) [inline doc/examples/streamline_formats.py]: functions

File: doc/examples/streamline_formats.py [hunk 18]
```diff
@@ -9,78 +9,223 @@
...
+the tractogram is the operation is computed only necessary.
```
Comment (Garyfallidis) [inline doc/examples/streamline_formats.py]: the tractogram is operating and compute only necessary transformations

File: doc/examples/streamline_formats.py [hunk 19]
```diff
@@ -9,78 +9,223 @@
...
+
+No matter the state, function such as saving or removing invalid coordinate
```
Comment (Garyfallidis) [inline doc/examples/streamline_formats.py]: functions... coordinates

File: doc/examples/streamline_formats.py [hunk 20]
```diff
@@ -9,78 +9,223 @@
...
+can be called safely and the transformation are handle internally if needed.
```
Comment (Garyfallidis) [inline doc/examples/streamline_formats.py]: transformations are handled internally when needed.

File: doc/examples/streamline_formats.py [hunk 21]
```diff
@@ -9,78 +9,223 @@
...
+affine, dimensions, voxel_sizes, voxel_order = cc_sft.space_attribute
+cc_density = density_map(cc_streamlines_vox, dimensions)
+laf_density = density_map(laf_streamlines_vox, dimensions)
+raf_density = density_map(raf_streamlines_vox, dimensions)
+lpt_density = density_map(lpt_streamlines_vox, dimensions)
+rpt_density = density_map(rpt_streamlines_vox, dimensions)
 
+"""
+Replacing streamlines is possible, but if the state was modified between
+operation such as this one is not recommended:
```
Comment (Garyfallidis) [inline doc/examples/streamline_formats.py]: operations

File: doc/examples/streamline_formats.py [hunk 22]
```diff
@@ -9,78 +9,223 @@
...
+-> cc_sft.streamlines = cc_streamlines_vox
+
+It is recommended to re-create a new StatefulTractogram object and
+explicitly specified in which space the streamlines are. Be careful to follow
```
Comment (Garyfallidis) [inline doc/examples/streamline_formats.py]: specify

File: doc/examples/streamline_formats.py [hunk 23]
```diff
@@ -9,78 +9,223 @@
...
+the order of operations.
+
+If the tractogram was from a Trk file with metadata, this will be lost.
+If you wish to keep metadata while manipulating the number or the order
+look the function StatefulTractogram.remove_invalid_streamlines() for more
```
Comment (Garyfallidis) [inline doc/examples/streamline_formats.py]: look at

**Comment by skoudoro** [comment]
Thank you @frheault, merging

**Pull request merged**